### PR TITLE
3.0 api docs

### PR DESCRIPTION
**Pull request by rictic**
Fixes https://github.com/Polymer/docs/issues/2379

Needs https://github.com/Polymer/tools/pull/220 in order to produce this output, though we could merge this before that lands if we needed to.

### CODE REVIEW

File: scripts/generate_api_docs_3.js [hunk 1]
```diff
@@ -0,0 +1,501 @@
+/**
```
Comment (justinfagnani) [inline scripts/generate_api_docs_3.js]: license

File: scripts/generate_api_docs_3.js [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+/** @typedef {format.Element|format.Class|format.Function|format.ElementMixin} SimpleFeature */
+/** @typedef {Map<string, {allFeatures: SimpleFeature[], analysis: format.Analysis}>} FeatureIndex */
+
+/** @param {format.Analysis} analysis */
+function indexFeaturesByFile(analysis) {
+  /** @type {FeatureIndex} */
+  const map = new Map();
+  /** @param {SimpleFeature} feature */
+  function index(feature, kind) {
+    const filename = getFilename(feature) || '???';
```
Comment (justinfagnani) [inline scripts/generate_api_docs_3.js]: warn?

File: scripts/generate_api_docs_3.js [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+<section anchor-id="${subsection}">
+  <h2>${subsection}</h2>
+
+  ${filenamesIn.map(fn => getFilenameDescription(fn, featureIndex)).join('\n\n')}
+</section>
+`
+}
+
+const hardcodedDescriptions = new Map([
+  ['lib/elements/array-selector.js', `
```
Comment (justinfagnani) [inline scripts/generate_api_docs_3.js]: What's the way forward here? Can you open issues in Analyzer and Polymer to add these to the source?

File: scripts/generate_api_docs_3.js [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+    throw new Error(`Got no features for filename ${filename}`);
+  }
+  const features = result.allFeatures;
+  const shortName = filename.split('/').slice(2).join('/');
+  let description;
+  if (features.length === 1) {
+    const [feature] = features;
+    description = feature.summary;
+  }
+  description = description || hardcodedDescriptions.get(filename) || `TODO(write me): ${filename}`;
```
Comment (justinfagnani) [inline scripts/generate_api_docs_3.js]: TODO or just warn?

Maybe these debug defaults should only be written with a flag on.

**Comment by rictic** [comment]
All comments addressed. https://github.com/Polymer/tools/issues/241 is the path forward for file overviews, and instead of producing ??? or TODO in the final output, it throws instead.

**Comment by arthurevans** [comment]
A few thoughts:

Current file overview is a little confusing. Especially weird to have the tiny import line be the first thing on the page, followed by a giant headline for the element/class. This might look better once the `@fileoverview` tag is implemented.

Current format doesn't seem to document symbols that are re-exported. For example, the `polymer-element Module` page doesn't show the `html` function.

My first thought was that it was weird to have the element/class embedded directly in the module page , but I guess the vast majority of our modules only export one thing. The challenge here is that if a single module exported two elements, or an element and some functions, it would be virtually impossible to figure out that the other items were there unless you knew to look for them.

I think this issue could be addressed with a quick-reference at the beginning of the file, so you have something like:

File overview description

import {} from '@polymer/polymer/polymer-element.js' 

Exports: PolymerElement, html, fooBar

Where each item in the Exports list is a link to the corresponding item.

But this might look weird if almost all of our files have only a single export. 

A TOC or some other navigation aid would serve the same purpose of orienting people in the file (in fact, it'd be even better for large elements/mixins with lots of props and methods). Maybe we can get someone with a design-y eye to look at it.

As far as the URLs go, the current URL fragments assume that each module is only exporting a single class/element. For example, if you had a single module that exported simple-dialog and fancy-dialog, each with an `open` method, the fragments for both would look identical:

my-dialogs#method-open

Not sure if I am borrowing trouble here. But I'd hate to rewrite all of the links to the API docs and then change the URLs, so let's discuss this assumption.

**Comment by arthurevans** [comment]
Opened https://github.com/Polymer/docs/issues/2541 to track suggestions above.

**Pull request merged**